### ISSUE TITLE
valid_target_types AttributeError from columns endpoint

### ISSUE THREAD

**Issue by seancolsen** [centerofci/mathesar#1855]
I got this API error from the columns endpoint totally randomly. I had not performed any DDL or DML mutations in quite a while and had been previously getting successful responses from the same endpoint. I was not making lots of requests in parallel. I was just doing local testing of the record selector, which only ever performs GET requests. After this one error response, I continued to get successful responses.

<details>
<summary>Traceback</summary>

```
Environment:


Request Method: GET
Request URL: http://localhost:8000/api/db/v0/tables/5/columns/?limit=500

Django Version: 3.1.14
Python Version: 3.9.9
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'rest_framework',
 'django_filters',
 'django_property_filter',
 'mathesar']
Installed Middleware:
['django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware']



Traceback (most recent call last):
  File "/code/mathesar/models/base.py", line 685, in __getattribute__
    return super().__getattribute__(name)

During handling of the above exception ('Column' object has no attribute 'valid_target_types'), another exception occurred:
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 826, in __getattr__
    return getattr(self.comparator, key)

The above exception ('Comparator' object has no attribute 'valid_target_types') was the direct cause of the following exception:
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/viewsets.py", line 125, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 466, in handle_exception
    response = exception_handler(exc, context)
  File "/code/mathesar/exception_handlers.py", line 55, in mathesar_exception_handler
    raise exc
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 506, in dispatch
    response = handler(request, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/mixins.py", line 43, in list
    return self.get_paginated_response(serializer.data)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 745, in data
    ret = super().data
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 246, in data
    self._data = self.to_representation(self.instance)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 663, in to_representation
    return [
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 664, in <listcomp>
    self.child.to_representation(item) for item in iterable
  File "/code/mathesar/api/serializers/columns.py", line 84, in to_representation
    representation = super().to_representation(instance)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 515, in to_representation
    ret[field.field_name] = field.to_representation(attribute)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/fields.py", line 1882, in to_representation
    return method(value)
  File "/code/mathesar/api/serializers/columns.py", line 204, in get_valid_target_types
    valid_target_types = column.valid_target_types
  File "/code/mathesar/models/base.py", line 691, in __getattribute__
    return getattr(self._sa_column, name)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 828, in __getattr__
    util.raise_(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception

Exception Type: AttributeError at /api/db/v0/tables/5/columns/
Exception Value: Neither 'Column' object nor 'Comparator' object has an attribute 'valid_target_types'
```

</details>

**Comment by seancolsen** [comment]
Just want to follow up that I've gotten this error about 5 times in the past day as I've been using the app. I'm submitting a lot of requests in my front end work right now, so in relative terms this bug rare, but I do think it's worth looking into.

**Comment by kgodey** [comment]
I updated the priority to High, so this should be on @silentninja's queue at some point.

**Comment by mathemancer** [comment]
I think this is a duplicate of #1791 . I'll reopen that one, and close this, since there's a bit more discussion there and it would be good to keep it consolidated. @seancolsen ping me if you disagree.

**Comment by seancolsen** [comment]
@mathemancer Perhaps it's worth highlighting the difference between this issue and #1791 that led me to report them separately... 

- In this issue, I hit the error and then am able to submit the exact same request again without getting the error. The error is stateless.
- In #1791, after I hit the error, all subsequent requests produced the same error until I restarted Docker. The error is stateful. 

If that distinction seems irrelevant to you, then feel free to leave this ticket closed.

**Comment by silentninja** [comment]
I agree with @seancolsen. This seems to be a different bug
- #1791 was happening due to the underlying column not existing on the user database, so the equivalent SqlAlchemy `Column` object could not be found. But the stacktrace for this bug is able to get the `_sa_column` and is failing at that point.
  ```
  File "/code/mathesar/models/base.py", line 691, in __getattribute__
    return getattr(self._sa_column, name)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 828, in __getattr__
    util.raise_(
    
   ```

**Comment by mathemancer** [comment]
Okay, I'm convinced it's distinct.

**Comment by mathemancer** [comment]
@seancolsen I can't reproduce this; sorry. To get me moving in spite of that:
- is this still happening?
- Are the error messages consistent when they happen? In particular, have you noticed whether
  ```python
      return getattr(self._sa_column, name)
  ```
  is always involved?
- Are there any other symptoms when it happens?
- Is it always under the `/tables/<id>/columns/` endpoint, and if so, is it always when the sequence of requests made upon loading a table page occur?

**Comment by seancolsen** [comment]
@mathemancer 

- > is this still happening?

    Yes, but it's rare. I probably hit it about once or twice per day. But I'm making at least a hundred such requests per day during local development.

- > Are the error messages consistent when they happen?

    Not sure. I'll start manually logging the error messages when I get them. Then I can compare them.

- > Are there any other symptoms when it happens?

    Nope.

- > Is it always under the `/tables/<id>/columns/` endpoint?

    Yes.

- > Is it always when the sequence of requests made upon loading a table page occur?

    Sort of -- I'm observing this error when loading the Record Selector (because that's what I'm working on). I may have also observed it when loading the Table Page, but I'm not sure. The sequence of requests _should_ (in theory) be the same for the two workflows. I suppose it's possible this error could somehow be limited to the Record Selector workflow, but that would surprise me. In both workflows, the front end makes requests in parallel to fetch the columns, constraints, and records. The requests resolve in an arbitrary order.

**Comment by seancolsen** [comment]
I'm not seeing this error anymore so I'm closing this ticket :crossed_fingers:
